src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1040 @@
+ *                                                                         *
+ * Source is provided to this software because we believe users have a     *
+ * right to know exactly what a program is going to do before they run it. *
+ * This also allows you to audit the software for security holes.          *
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <wchar.h>
+#include <string.h>
+
+/* Offset retrieval */
+#include "vmi_win_gui_offsets.h"
+#include "vmi_win_gui_atoms.h"
+#include "vmi_win_gui_utils.h"
+#include "vmi_win_gui_filter_definitions.h"
+
+#include "vmi_win_gui_parser.h"
+
+struct rect* get_visible_rect_from_bitmask(
+    char* map, size_t n, int scanline, struct rect* r)
+{
+    struct rect* result = NULL;
+
+    size_t byte;
+    int bit_idx;
+    unsigned int bit;
+    int x0 = -1, x1 = -1, y0 = -1, y1 = -1;
+    bool is_y_hole = false;
+    bool is_x_hole = false;
+
+    for (int y = r->y0; y < r->y1; y++)
+    {
+        for (int x = r->x0; x < r->x1; x++)
+        {
+            byte = (y * scanline + x)/8;
+
+            /* Parts of a wnd can be outside of the desktop's frame */
+            if (byte >= n || x < 0)
```

Comment:
can `x < 0` be ever true?

Reply:
Yes, I double checked that. Rare cases, where a window and its relevant button protrudes the left frame edge.